High level integration tests

elliotcmorris opened the pull request:
Closes #12 

Add a set of high level integration tests, and data to go with them. All they do is load a USD stage, then verify that it manages to find the properties of the branch prims.

These tests assert the basic workflow of loading a USD document with resolvable references in it, as well as that the OpenAssetIO resolver dosen't override or interfere with USD standard resolution mechanisms.

For any test dataset that expects OpenAssetIO to resolve a LocatableContent trait, a BAL format library has been included with the dataset to document that path strings that should be returns, although no effort has been made to integrate OpenAssetIO or BAL, the formatting choice is purely speculative.

elliotcmorris committed:
[Test] Add integration test datasets.

Add datasets for USD integration tests, high level USD scenes that
reference paths via search-path, non-search-path, and assetized
references.

These tests are so we can verify implementation works as we expect
as we implement resolver mechanisms.

Signed-off-by: Elliot Morris <elliot.morris@foundry.com>

foundrytom commented on tests/resources/integration_test_data/assetized_child_ref_non_assetized_grandchild/parking_lot.usd:
```diff
@@ -0,0 +1,11 @@
+def "ParkingLot"
+{
+    def "ParkingLot_Floor_1" (
+        references = @floor@</ParkingLot_Floor>
```
[fix] All the assetized refs would need to be `@bal:///floor@...` otherwise BAL will ignore them.

elliotcmorris replied:
https://github.com/OpenAssetIO/usdOpenAssetIOResolver/pull/13/commits/6a194ba8a1a6a29c67d4d722e3d1c8f224b45c71

Arguably i'm not prescribing BAL as neccesarily the manager for this dataset ... but that argument is bunk because that's totally my intent.

(Also fixed a floor data file that i'd deleted all the content from when debugging to see if it'd load, lots of debris from my battle with the resolver context)

foundrytom commented on tests/resources/integration_test_data/non_assetized_child_ref_assetized_grandchild/parking_lot.usd:
```diff
@@ -0,0 +1,11 @@
+def "ParkingLot"
+{
+    def "ParkingLot_Floor_1" (
+        references = @floor.usd@</ParkingLot_Floor>
```
[fix] `floor1.usd`?

elliotcmorris replied:
https://github.com/OpenAssetIO/usdOpenAssetIOResolver/pull/13/commits/f7c34dd956000513d60f586899e4c716eb268789

foundrytom commented on tests/test_resolver.py:
```diff
@@ -10,9 +10,33 @@
+def build_search_path_context(path_relative_from_file):
+    script_dir = os.path.realpath(os.path.dirname(__file__))
+    full_path = os.path.join(script_dir, path_relative_from_file)
+    return Ar.DefaultResolverContext([path_relative_from_file])
```
[fix] Should this be `full_path`?

```suggestion
    return Ar.DefaultResolverContext([full_path])
```

elliotcmorris replied:
Ah, yes, another left over symptom of all my search path hacking. Annoyingly (or not, depending on your outlook,) USD will append the working directory if it detects the path is not absolute, so this continues to "work". But as you mentioned before, this would fall over if you ran the tests from an unexpected working directory.

elliotcmorris replied:
https://github.com/OpenAssetIO/usdOpenAssetIOResolver/pull/13/commits/ef46ffb6ca0fda46223aa4fb1e6aff175b7d41de

foundrytom commented on tests/test_resolver.py:
```diff
@@ -26,3 +50,99 @@ def test_open_stage_and_logging(capfd):
+        "resources/integration_test_data/resolver_has_no_effect_with_search_path/search_path_root"
+    )
+    stage = Usd.Stage.Open(
+        "resources/integration_test_data/resolver_has_no_effect_with_no_search_path/parking_lot.usd",
```
[fix]
```suggestion
        "resources/integration_test_data/resolver_has_no_effect_with_search_path/parking_lot.usd",
```

elliotcmorris replied:
https://github.com/OpenAssetIO/usdOpenAssetIOResolver/pull/13/commits/b1874c01e2323038189d17f64c2c55e9a01e2638

foundrytom commented on tests/test_resolver.py:
```diff
@@ -26,3 +50,99 @@ def test_open_stage_and_logging(capfd):
+# As all the data tends to follow the same form, convenience method
+# to avoid repeating myself
+def assert_parking_lot_structure(usd_stage):
+    floor = usd_stage.GetPrimAtPath("/ParkingLot/ParkingLot_Floor_1")
```
[fix] I think itd be worth being explicit and adding `assert floor.IsValid()`, etc. So we're not implicitly relying on the specifics of the erst of the test.

elliotcmorris replied:
https://github.com/OpenAssetIO/usdOpenAssetIOResolver/pull/13/commits/3dbf21674ddde2760055681f4de77e64395d3b49

foundrytom commented on tests/test_resolver.py:
```diff
@@ -26,3 +50,99 @@ def test_open_stage_and_logging(capfd):
+# to avoid repeating myself
+def assert_parking_lot_structure(usd_stage):
+    floor = usd_stage.GetPrimAtPath("/ParkingLot/ParkingLot_Floor_1")
+    car1 = usd_stage.GetPrimAtPath("/ParkingLot/ParkingLot_Floor_1/Car1")
+    car2 = floor.GetChild("Car2")
```
[question] Any particular reason to split the lookup approach? The second form seems easier to maintain.

elliotcmorris replied:
It's a vague notion, but I thought to do this for the parent assetized/child non assetized test, just in case USD does some under-the-hood things where it can bypass resolving the intermediate file. It's just a small flourish really, will remove if you think it's any maintenance burden.

foundrytom replied:
Id lean towards the simpler/consistent version

elliotcmorris replied:
https://github.com/OpenAssetIO/usdOpenAssetIOResolver/pull/13/commits/2aad5788a6802d7074bc48fc8698d104ce777038

elliotcmorris replied:
Possibly based on Davids comments below we'll remove the duplicate car entries now, since they're accessed as exact duplicates.

foundrytom replied:
I kind of like there being two as it means were exercising multiple resolves of the same path

foundrytom commented on tests/test_resolver.py:
```diff
@@ -26,3 +50,99 @@ def test_open_stage_and_logging(capfd):
+    assert_parking_lot_structure(stage)
+
+
+# Given a USD document that recursively references multiple other USD
```
[fix] I struggled to grok this for a bit Can we use 'assetized' here instead of just 'resolves/references' - as resolve is a USD term ie. everything is being resolved by `Ar` even if its a file path. Might be worth updating the folder names to be consistent too?

I mention this not to nit-pick, but this wording is defining the expected behaviour so needs to be unambiguous in the future about path based references and assetized ones. How you have it for the child/grandchild scenarios seems perfect.

elliotcmorris replied:
I'm not sure if this is what you're getting at, to my eye these are mostly equivilant so I'm liable to have misinterpreted you. Let me know if this is what you had in mind.

https://github.com/OpenAssetIO/usdOpenAssetIOResolver/pull/13/commits/22dce9219328fd7f3b375d57ca189a8aa66225b2

(fixup that moves the data files to the new name)
https://github.com/OpenAssetIO/usdOpenAssetIOResolver/pull/13/commits/d91d50a3b0a659d8b3bae935a2673c88283e7ded

foundrytom replied:
Perfect, thanks, it was the missing 'assetized' bit - as all things are resolved...

foundrytom reviewed (changes requested):
Nice pragmatic coverage 👍.

As a side note - we generally have been using the when/then form for test names. Be good to align on which we prefer. I kind of like this approach in many ways as it moves the when/then into a comment which has more space than <99 chars.

feltech commented on tests/resources/integration_test_data/recursive_resolve_to_search_path/bal_library.json:
```diff
@@ -0,0 +1,26 @@
+        {
+          "traits": {
+            "openassetio-mediacreation:content.LocatableContent": {
+              "location": "cars/car.usd"
```
[question] There is also the special case of `./`/ `../` prefixes, which disable search path lookups.

Based on previous discussions, perhaps we can be a bit fuzzy about that, though, since we're then essentially testing `ArDefaultResolver`. 

Perhaps we can be convinced that if search path behaviour is working, then we've successfully passed through the inputs to `ArDefaultResolver`, and don't need to worry about `./`?  Not sure about that though.

foundrytom replied:
I think this is plenty for now. In terms of "passing through to the resolver", Id definitely raise the point from before of - we need a way to verify this that isn't entirely coupled to its current behaviour. Perhaps integration tests aren't the best way to assert that.

elliotcmorris replied:
I don't think we need to treat `file.usd` and `./file.usd` any differently. They trigger the same call path via [this](https://github.com/PixarAnimationStudios/USD/blob/5c5ebddff35012461a2b0ba773c47f05730cbab4/pxr/usd/ar/defaultResolver.cpp#L65) call, it's a bit convoluted, but this ends up checking whether the path starts in "./" or "../", or that it dosen't start in "/", giving the same result, albeit via two different branches. [TfsIsRelativePath here](https://github.com/PixarAnimationStudios/USD/blob/857ffda41f4f1553fe1019ac7c7b4f08c233a7bb/pxr/base/tf/pathUtils.cpp#L290)

feltech commented on tests/test_resolver.py:
```diff
@@ -10,9 +10,33 @@
+
+    assert (
+        plugin is not None
+    ), "OpenAssetIO is not configured as a resolver. Set `PXR_PLUGINPATH_NAME` to `plugInfo.json`"
```
[consider][minor] Not sure about the wording here. 

Checking if the plugin exists doesn't confirm that it's configured as a resolver, just that its loaded. 

Also, surely in most cases, simply setting `PXR_PLUGINPATH_NAME` to `plugInfo.json` wouldn't work.

Perhaps "The usdOpenAssetIOResolver plugin is not loaded, please check the PXR_PLUGINPATH_NAME environment variable"?

elliotcmorris replied:
https://github.com/OpenAssetIO/usdOpenAssetIOResolver/pull/13/commits/a7438e38999cdf14cfd0b1d8c6db55f1cb0affe1

feltech commented on tests/test_resolver.py:
```diff
@@ -26,3 +50,99 @@ def test_open_stage_and_logging(capfd):
+def test_openassetio_resolver_has_no_effect_with_no_search_path():
+
+    stage = Usd.Stage.Open(
+        "resources/integration_test_data/resolver_has_no_effect_with_no_search_path/parking_lot.usd"
```
[consider] This is a little brittle, since it assumes the cwd is the `tests` directory.

(first time I ran the tests locally on the command-line they failed because I wasn't in the correct directory)

foundrytom replied:
I'd be happy with a caveat for now that is clearly stated in the docs, something we can improve once we have things functionally working?

elliotcmorris replied:
I thought it was worth it, as we've gone to the same effort for the search path. Uses the same mechanism.

https://github.com/OpenAssetIO/usdOpenAssetIOResolver/pull/13/commits/fd8f8278784b757f830d5457c4b4c0e1a5dd7fe5

feltech commented:
> As a side note - we generally have been using the when/then form for test names. Be good to align on which we prefer. I kind of like this approach in many ways as it moves the when/then into a comment which has more space than <99 chars.

Elaborating in comments is a good idea, I like that.  But the test function names are what is shown in the `pytest` output, so I think are worth keeping as descriptive as possible.

As an aside, I'm pretty sure the `unittest` runner will print docstrings rather than method names, if available.  But pytest doesn't.  A quick google finds some pytest plugins though, which may be worth a look.

feltech commented on tests/resources/integration_test_data/assetized_child_ref_non_assetized_grandchild/floor1.usd:
```diff
@@ -0,0 +1,16 @@
+        references = @car.usd@</Car>
+    )
+    {
+        color3f color = (0, 0, 1)
```
[consider] We could simplify and remove these overrides, since it doesn't look like we need these for testing. I found myself scanning the code for asserts on the correct value of `color`, which is never done, so is a bit misleading.

We could use the `color` override to assert that the expected child was chosen (e.g. to assert that local relative has precedence over search-path relative), but I think that would be testing `ArDefaultResolver` and not affected by our integration(?).

feltech replied:
I think the point is more general. It would be good to simplify the dataset to the absolute bare minimum required for the test(s).

elliotcmorris replied:
I think we can cul us down to only having a single car on the floor, now that we're not accessing the cars in different ways, (the multiple cars was to test caching initially anyhow, so you're right they're not appropriate,) there's also no need for the duplicate and override properties yeah.

foundrytom replied:
Id like to keep the two cars for the reasons elliot described, but agree that colors are tangential at this point.

elliotcmorris replied:
https://github.com/OpenAssetIO/usdOpenAssetIOResolver/pull/13/commits/15b8c3684dffefe00a8d83aa05f309407a86c82c

https://github.com/OpenAssetIO/usdOpenAssetIOResolver/pull/13/commits/3b5ae49a4f1dc1d2c2adbd4512fd83aba23646b6

Removed redundant properties.

feltech commented on tests/test_resolver.py:
```diff
@@ -26,3 +50,99 @@ def test_open_stage_and_logging(capfd):
+@pytest.mark.xfail(reason="OpenAssetIO not integrated yet")
+def test_non_assetized_child_ref_assetized_grandchild():
+    stage = Usd.Stage.Open(
+        "resources/integration_test_data/non_assetized_child_ref_assetized_grandchild/parking_lot.usd"
```
[fix] We're gonna have to do some `pylint`ing and `black`ing, to avoid having to fixup once the CI linting PR is merged.

elliotcmorris replied:
https://github.com/OpenAssetIO/usdOpenAssetIOResolver/pull/13/commits/70e43c109ea708191783e554ad0b07b0af461653

pylint and black changes. Did it manually so possible the merged lint changes could still catch some things.

feltech commented on tests/test_resolver.py:
```diff
@@ -26,3 +50,99 @@ def test_open_stage_and_logging(capfd):
+
+
+# Given a USD document that references an asset via a relative search-path
+# based file path, then the asset is resolved the the file path as expected.
```
[fix] 
```suggestion
# based file path, then the asset is resolved to the file path as expected.
```

feltech replied:
Also exceeds 72 char.

elliotcmorris replied:
https://github.com/OpenAssetIO/usdOpenAssetIOResolver/pull/13/commits/c2d4bea59cd3a76db852031a1eedac579cc4f683

elliotcmorris commented:
> As a side note - we generally have been using the when/then form for test names. Be good to align on which we prefer. I kind of like this approach in many ways as it moves the when/then into a comment which has more space than <99 chars.

> Elaborating in comments is a good idea, I like that.  But the test function names are what is shown in the `pytest` output, so I think are worth keeping as descriptive as possible.

I think you'd get some really absurdly long method names if you were to do that in these cases, I did try initially. I think the names are pretty descriptive as they are, just shortened from the full given;and;then framing. That would also propagate to give some really long data folder names to match, (although arguably that's not a neccesary connection)

feltech commented on tests/test_resolver.py:
```diff
@@ -10,9 +10,33 @@
+
+# Verify OpenAssetIO configured as the AR resolver.
+@pytest.fixture(autouse=True)
+def openassetio_configured():
```
[consider][minor] We do try to put utility functions at the bottom of the file, with the general principle that we should be able to read source code from high level to low level as we scroll down.

This style has been our preference so far (though not always obeyed).  Happy to discuss if you have a counter opinion.

elliotcmorris replied:
Aye i'm not doing this intentionally, just habit, the idea being each unit of code maintains maximum local cohesion, so you can almost treat sections of the file as seperate, leaving yourself open to easily factor out to hard boundaries (ie, new files) later when the scope of the sections justify it.

Gonna take some adjusting to scroll _down_ to find utility functions, :P

Will conform to standards.

foundrytom replied:
To me there is an argument that utilities should be non-local as they may well be shared, and that makes it clear. ie, if you had a large file of tests, you could accidentally re-implement something several times as you hadn't seen it, as it was local to some specific use?

elliotcmorris replied:
Aye, very fair and true. A true module _extremist_ might argue that duplicate utility implementations is fine, and even preferred, in order to keep cross-module dependencies mapped exclusively to data communication, and not to implementational detail.

There's a section in _Clean Code_ about this, page 140 if you're interested, although it speaks exclusively about module/class locality, my extension of the reasoning to file scope is personal and tenuous, and not something I'd wanna stand by :P.

elliotcmorris replied:
You know, reading that section of the book again, I find it hard to imagine how I went from there to, "Things should be really close together in the source file,) you could easily take a fully contrary interpretation. Ach well, no matter.

elliotcmorris replied:
https://github.com/OpenAssetIO/usdOpenAssetIOResolver/pull/13/commits/b22d39f05c43d3a7021a51aafef71d110acad39d

Util funcs at the bottom

feltech commented:
> > As a side note - we generally have been using the when/then form for test names. Be good to align on which we prefer. I kind of like this approach in many ways as it moves the when/then into a comment which has more space than <99 chars.
> 
> > Elaborating in comments is a good idea, I like that.  But the test function names are what is shown in the `pytest` output, so I think are worth keeping as descriptive as possible.
> 
> I think you'd get some really absurdly long method names if you were to do that in these cases, I did try initially. I think the names are pretty descriptive as they are, just shortened from the full given;and;then framing. That would also propagate to give some really long data folder names to match, (although arguably that's not a neccesary connection)

To be fair, we also say that `when`/`then` style is only really useful in cases where there are multiple branches. In these cases, you could possibly write a `when` (though it would indeed be wordy), but the `then` is always the same so is a bit redundant. 

I'm happy in this case to have the test case name essentially describe the dataset. If a test fails, then that would make it pretty easy to tell what functionality was broken at a glance.

foundrytom reviewed (approved):
Looking good in terms of my comments, thanks @elliotcmorris

foundrytom commented on tests/test_resolver.py:
```diff
@@ -26,3 +33,145 @@ def test_open_stage_and_logging(capfd):
+    script_dir = os.path.realpath(os.path.dirname(__file__))
+    full_path = os.path.join(script_dir, path_relative_from_file)
+
+    print(full_path)
```
[fix] Should this still be here?

elliotcmorris replied:
_sigh_ ... no.

elliotcmorris replied:
https://github.com/OpenAssetIO/usdOpenAssetIOResolver/pull/13/commits/ec8d357b41cc84f6fbf39e00f13298cc17255b13

foundrytom commented on tests/test_resolver.py:
```diff
@@ -26,3 +33,145 @@ def test_open_stage_and_logging(capfd):
+    return Ar.DefaultResolverContext([full_path])
+
+
+# Open the stage
+# Done this way so that you can run the test from any directory,
+# otherwise the working directory will impact the file loading.
```
As an aside (we can tweak later), in Python, we'd normally write these as docstrings, so theyre available at runtime, and you dont need to mollify the linter.

feltech replied:
+1 to this, I even think this warrants a [consider].

elliotcmorris replied:
Whilst this would be trivial, will do later anyhow.

foundrytom commented on tests/resources/integration_test_data/assetized_child_ref_non_assetized_grandchild/bal_library.json:
```diff
@@ -0,0 +1,15 @@
+        {
+          "traits": {
+            "openassetio-mediacreation:content.LocatableContent": {
+              "location": "floor1.usd"
```
[fix] These need to be `file:///` URLs

elliotcmorris replied:
[a38c6af](https://github.com/OpenAssetIO/usdOpenAssetIOResolver/pull/13/commits/a38c6af2977c4e2ba0a3b0c98fb37539bbad6286)

feltech reviewed (approved):
Looking good.

elliotcmorris commented:
Changes based on pivots discussed offline, not longer allowing relative file path from resolver.

[a682351](https://github.com/OpenAssetIO/usdOpenAssetIOResolver/pull/13/commits/a682351b6537a6f9e0616d0e2086d2dfe2f69723)

[a38c6af](https://github.com/OpenAssetIO/usdOpenAssetIOResolver/pull/13/commits/a38c6af2977c4e2ba0a3b0c98fb37539bbad6286)

elliotcmorris committed:
[Test] High level integration tests

These tests are so we can verify implementation works as we expect
as we implement resolver mechanisms.

Signed-off-by: Elliot Morris <elliot.morris@foundry.com>

feltech commented:
> Changes based on pivots discussed offline, not longer allowing relative file path from resolver.
> 
> [a682351](https://github.com/OpenAssetIO/usdOpenAssetIOResolver/pull/13/commits/a682351b6537a6f9e0616d0e2086d2dfe2f69723)
> 
> [a38c6af](https://github.com/OpenAssetIO/usdOpenAssetIOResolver/pull/13/commits/a38c6af2977c4e2ba0a3b0c98fb37539bbad6286)

Unfortunately GitHub aggressively prunes branches after a force-push, so those commits are no longer available.  Think I can compare locally though.

elliotcmorris commented:
> Unfortunately GitHub aggressively prunes branches after a force-push, so those commits are no longer available. Think I can compare locally though.

Yeah sorry David, in a bit of a race to hit the button on this one, thought those commits could be taken as read.

feltech reviewed (approved):
Latest version looks fine. Done well only having one pylint issue to deal with!
